src/Sarif/Visitors/SarifCurrentToVersionOneVisitor.cs

```diff
@@ -76,6 +122,32 @@ internal AnnotationVersionOne CreateAnnotation(Region v2Region)
             return annotation;
         }
 
+        internal CodeFlowVersionOne CreateCodeFlow(CodeFlow v2CodeFlow)
+        {
+            CodeFlowVersionOne codeFlow = null;
+
+            if (v2CodeFlow != null)
+            {
+                codeFlow = new CodeFlowVersionOne
+                {
+                    Message = v2CodeFlow.Message?.Text,
+                    Properties = v2CodeFlow.Properties
+                };
+
+                if (v2CodeFlow.ThreadFlows?.Count > 0)
+                {
+                    codeFlow.Locations = new List<AnnotatedCodeLocationVersionOne>();
+
+                    foreach (ThreadFlow tf in v2CodeFlow.ThreadFlows)
+                    {
+                        (codeFlow.Locations as List<AnnotatedCodeLocationVersionOne>).AddRange(CreateAnnotatedCodeLocation(tf));
```

Comment:
Unfortunately it's not this easy; in fact, it's going to be quite difficult. Please read the following carefully.

In v1, a `codeFlow` was described by a single list of `annotatedCodeLocation`s, in chronological order. To represent the fact that there might be multiple threads of execution, `annotatedCodeLocation` had a `threadId` property:

> ... whose value is an integer which identifies the thread that was executing when the execution of a code flow reached the location specified by this object. 

For example:
```
{                                           # A codeFlow object
  "locations": [
    {
      "step": 1,
      "threadId": 42
    },
    {
      "step": 2,
      "threadId": 54
    },
    {
      "step": 3,
      "threadId": 42
    }
  ]
}
```
Step 1 and Step 3 occur on thread 42, Step 2 occurs on thread 54, and Step 2 occurs after Step 1 and before Step 3.

In v2, we decided to represent each thread separately, so a `codeFlow` contains an array of `threadFlow`s. To represent the inter-thread ordering information of the original design, we added a property `threadFlowLocation.executionOrder`.

For example:
```
{                                           # A codeFlow object
  "threadFlows": [
    {                                       # A threadFlow object
      "id": "42",
      "locations": [
        {
          "step": 1
          "executionOrder": 1
        },
        {
          "step": 2,
          "executionOrder": 3
        }
      ]
    },
    {                                       # Another threadFlow object
      "id": 54,
      "locations": [
        {
          "step": 1
          "executionOrder": 2
        }
      ]
    }
  ]
}
```

So when you translate from v1 to v2, you have do something like this:
```
executionOrder = 0
for each location in the v1 codeFlow
  if we have not seen the location's thread id before (and this includes the case where it's absent aka 0)
    create a new thread flow
    threadFlow.id = stringified location.threadId (culture-invariant)
 add the location to the appropriate threadFlow, setting its executionOrder to the next available value
```
I'll leave it to you to figure out the v2 to v1 logic. Be aware that `executionOrder` is _optional_.

This needs to be thoroughly unit tested. If you're going to stick to the current unit test strategy of converting an entire file, make sure that you use the absolute minimum file -- no extra properties. Just the ones you need for each test case.

Reply:
Larry, for v1 -> v2, how should step be handled? Should I always set a value?

Reply:
If you need to set it at all (it is optional), you should set it on all `threadFlowLocation`s. Each `threadFlow` has its own sequence (1, 2, 3...).

@michaelcfanning The spec is not clear on this point, but it should be: if anybody has a step, everybody should have one, because the purpose is for users to be able to look at it and say "look what happened at step 4!".

I filed oasis-tcs/sarif-spec#203 on that.

Reply:
If it's not clear to you whether you need to set it in a given `codeFlow`, just go ahead and set it always.

Reply:
Given that executionOrder isn't required, how am I to handle sequence if it's missing? Just go with the list item order? And then there's the situation where some TFLs have a value but some don't.

Reply:
Yes, go with list order.

Example: Given this v2 code flow:
```
thread 1
    step 1
    step 2
    step 3: executionOrder = 1
    step 4:

thread 2
    step 1
    step 2 executionOrder = 2
```

This means: "thread 2 step 2 occurs after thread 1 step 3" (and that's _all_ it means!).

Given that, there are many valid conversions to a v1 thread flow. One possible translation is:
```
step 1: threadId = 1
step 2: threadId = 1
step 3: threadId = 1
step 4: threadId = 2 <== this was originally step 1 in thread 2!
step 5: threadId = 2 <== this was originally step 2 in thread 2
step 6: threadId = 1 <== this was originally step 4 in thread 1
```

Reply:
@michaelcfanning You see how tricky this is. It would be reasonable to just abandon v2 => v1 down-versioning at this point. The fact that it's "almost finished" isn't a good enough reason to finish it. That's the "sunk cost fallacy". The time you'd spend on that would be better spent _removing_ v2 => v1 conversion and its tests. IMNSHO.

Reply:
An important consideration is the iffy quality of the v1 log that we're able to produce. The v2 schema is a lot more sophisticated, there are few required properties, and there are many significant differences, so we end up either guessing or dropping data at a number of points during transformation.
And my current codeFlow transformer architecture won't work for the above so I'll spend extra time refactoring.

Reply:
@michaelcfanning Thoughts here Michael?